src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -367,9 +370,37 @@ public static string UrlEncode(string value)
             if (string.IsNullOrEmpty(value))
                 return value;
 
-            byte[] bytes = Encoding.UTF8.GetBytes(value);
-            byte[] encodedBytes = UrlEncode(bytes, 0, bytes.Length, false /* alwaysCreateNewReturnValue */);
-            return Encoding.UTF8.GetString(encodedBytes, 0, encodedBytes.Length);
+            int cSafe = 0;
+            int cSpaces = 0;
+            for (int i = 0; i < value.Length; i++)
+            {
+                char ch = value[i];
+                if (IsUrlSafeChar(ch))
+                {
+                    cSafe++;
+                }
+                else if (ch == ' ')
+                {
+                    cSpaces++;
+                }
+            }
+
+            if (cSafe == value.Length)
+            {
+                // Nothing to expand
+                return value;
+            }
+
+            int unsafeByteCount = Encoding.UTF8.GetByteCount(value) - cSafe - cSpaces;
+            int byteIndex = unsafeByteCount * 2;
+            
+            // For perf, we avoid allocating a new array for the encoded bytes, and
+            // instead encode in place
```

Comment:
I think it'd be worth expanding on this comment, e.g.
"Instead of allocating one array of length ```Encoding.UTF8.GetByteCount(value)``` to store the UTF8 encoded bytes and then a second array of length ```cSafe + cSpaces + 3 * (Encoding.UTF8.GetByteCount(value) - cSafe - cSpaces)``` to store the URL encoded UTF8 bytes, we allocate a single array of length ```cSafe + cSpaces + Encoding.UTF8.GetByteCount(value) * 3```, saving ```Encoding.UTF8.GetByteCount(value) - 3*(cSafe +cSpaces)``` bytes allocated.  We encode the UTF8 to the end of this array, and then URL encode to the beginning of the array."
or something like that.  If you decide to include a comment like this, including the math, please double-check my math :smile: